File: engine.py
class Engine:

    def __init__(self, data, op=(''), children=()) -> None:
        self.data = data
        self.grad = 0
        self._op = op
        self._prev = children
        
    def __repr__(self) -> str:
        return f"{self.data}, {self.grad}"
    
    def __add__(self, other):
        out = Engine(self.data + other.data, '+', (self, other))

        def _backward():
            self.grad += out.grad
            other.grad += out.grad
        out.backward = _backward

        print(out)
        return out

    def __mul__(self, other):
        out = Engine(self.data * other.data, '*', (self, other))

        def _backward():
            self.grad += other.data * out.grad
            other.grad += self.data * out.grad
        out.backward = _backward

        return out
    
    def __sub__(self, other):
        out = Engine(self.data - other.data, '-', (self, other))

        def _backward():
            self.grad += out.grad
            other.grad -= out.grad
        out.backward = _backward

        return out

    def __truediv__(self, other):     
        out = Engine(self.data / other.data, '/', (self, other))
    
        def _backward():
            self.grad += out.grad / other.data
            other.grad -= out.grad * self.data / other.data**2
        out.backward = _backward

        return out

    def __pow__(self, other):     
        out = Engine(self.data ** other.data, '**', (self, other))
    
        def _backward():
            self.grad += (other * self.data**(other-1)) * out.grad
        out.backward = _backward

        return out

    def relu(self):
        out = Engine(0 if self.data < 0 else self.data, (self,), 'ReLU')

        def _backward():
            self.grad += (out.data > 0) * out.grad
        out._backward = _backward

        return out

    def backward(self):
        topo = []
        visited = set()
        def build_topo(v):
            if v not in visited:
                visited.add(v)
                for child in v._prev:
                    build_topo(child)
                topo.append(v)
        build_topo(self)

        self.grad = 1
        for v in reversed(topo):
            v._backward()
        
    def __neg__(self): 
        return self * -1

    def __radd__(self, other): 
        return self + other

    def __sub__(self, other): 
        return self + (-other)

    def __rsub__(self, other):
        return other + (-self)

    def __rmul__(self, other): 
        return self * other

    def __truediv__(self, other): 
        return self * other**-1

    def __rtruediv__(self, other):
        return other * self**-1
    

File: cpp_code/engine.cpp
#include "engine.h"
#include <cmath>
#include <algorithm>
#include <iostream>
#include <sstream>

Engine::Engine(double data, const std::string& op, const std::vector<EnginePtr>& children)
    : data(data), grad(0), _op(op), _prev(children), _backward([](){}) {
}

EnginePtr makeEngine(double data, const std::string& op, const std::vector<EnginePtr>& children) {
    return std::make_shared<Engine>(data, op, children);
}

// Addition
EnginePtr operator+(const EnginePtr& a, const EnginePtr& b) {
    auto out = makeEngine(a->data + b->data, "+", {a, b});
    
    out->_backward = [out, a, b]() {
        a->grad += out->grad;
        b->grad += out->grad;
    };
    
    return out;
}

// Multiplication
EnginePtr operator*(const EnginePtr& a, const EnginePtr& b) {
    auto out = makeEngine(a->data * b->data, "*", {a, b});
    
    out->_backward = [out, a, b]() {
        a->grad += b->data * out->grad;
        b->grad += a->data * out->grad;
    };
    
    return out;
}

// Subtraction 
EnginePtr operator-(const EnginePtr& a, const EnginePtr& b) {
    auto out = makeEngine(a->data - b->data, "-", {a, b});
    
    out->_backward = [out, a, b]() {
        a->grad += out->grad;
        b->grad -= out->grad;
    };
    
    return out;
}

// Division 
EnginePtr operator/(const EnginePtr& a, const EnginePtr& b) {
    auto out = makeEngine(a->data / b->data, "/", {a, b});
    
    out->_backward = [out, a, b]() {
        a->grad += out->grad / b->data;
        b->grad -= out->grad * a->data / std::pow(b->data, 2);
    };
    
    return out;
}

// Power
EnginePtr pow(const EnginePtr& a, const EnginePtr& b) {
    auto out = makeEngine(std::pow(a->data, b->data), "**", {a, b});
    
    out->_backward = [out, a, b]() {
        a->grad += b->data * std::pow(a->data, b->data - 1) * out->grad;
        // Note: for the exponent gradient, we'd need logarithm, not implementing for simplicity
    };
    
    return out;
}

// ReLU
EnginePtr Engine::relu() {
    double relu_val = this->data < 0 ? 0 : this->data;
    auto out = makeEngine(relu_val, "ReLU", {this->getPtr()});
    
    out->_backward = [out, this]() {
        this->grad += (out->data > 0) ? out->grad : 0;
    };
    
    return out;
}

// Unary negation
EnginePtr operator-(const EnginePtr& a) {
    auto minus_one = makeEngine(-1);
    return minus_one * a;
}

// Backpropagation
void Engine::backward() {
    // Topological sort
    std::vector<EnginePtr> topo;
    std::set<EnginePtr> visited;
    
    std::function<void(const EnginePtr&)> build_topo = [&](const EnginePtr& v) {
        if (visited.find(v) == visited.end()) {
            visited.insert(v);
            for (const auto& child : v->_prev) {
                build_topo(child);
            }
            topo.push_back(v);
        }
    };
    
    // topological order
    build_topo(this->getPtr());
    
    // Set the gradient of the final output to 1
    this->grad = 1.0;
    
    // backpropagation in reverse topological order
    for (auto it = topo.rbegin(); it != topo.rend(); ++it) {
        (*it)->_backward();
    }
}

std::string Engine::toString() const {
    std::stringstream ss;
    ss << "data: " << data << ", grad: " << grad;
    return ss.str();
}
